irobot_create_toolbox/src/mock_publisher.cpp

```diff
@@ -143,6 +179,65 @@ MockPublisher::MockPublisher()
   battery_state_msg_.percentage = 1;
   // Set battery state header
   battery_state_msg_.header.frame_id = "base_link";
+  battery_state_msg_.capacity = 2.046;
+  battery_state_msg_.design_capacity = 2.046;
+  battery_state_msg_.present = true;
+  battery_state_msg_.temperature = 27.0;
+}
+
+double MockPublisher::get_docked_charge_percentage(const rclcpp::Time & at_time)
+{
+  const std::lock_guard<std::mutex> lock(battery_charge_timings_mutex_);
+  rclcpp::Duration time_docked = at_time - transitioned_to_docked_;
+  // current charge is last charged + charging rate * time
+  double docked_charge = last_docked_charge_percentage_ +
+    (time_docked.seconds() * charge_rate_percent_per_second_);
+  if (docked_charge > 1.0) {
+    docked_charge = 1.0;
+  }
+  return docked_charge;
```

Comment:
You can probably declare `docked_charge` amd `time_docked` as `const` above.

```suggestion
  return std::min(docked_charge, 1.0);
```